xoinvader/collision.py

```diff
@@ -1,49 +1,188 @@
+                Collider.__manager__ = None
+            Collider.__manager__ = weakref.ref(self, null_manager)
+        self._colliders = []
+        self._collisions = {}
+
+    def _add(self, collider):
+        self._colliders.append(collider)
+
+    def update(self):
+        """Detect and process all collisions."""
+        for pair in self._collisions:
+            c_a = [n for n in self._colliders if n.type == pair.a]
+            c_b = [n for n in self._colliders if n.type == pair.b]
+            for cldr_a in c_a:
+                for cldr_b in c_b:
+                    maybe_coll_rect = self.check_collision(cldr_a, cldr_b)
+                    if maybe_coll_rect:
+                        callbacks = self._collisions[pair]
+                        for cb in callbacks:
+                            if cb:
+                                cb(maybe_coll_rect)
+
+    def check_collision(self, col1, col2):
+        """Check collisions between two colliders.
+
+        Returns `None` if no collision occured, or returns rectangle of
+        overlapping region between collider maps.
+
+        :param col1: first collider
+        :type col1: :class:`Collider`
+        :param col2: second collider
+        :type col2: :class:`Collider`
+        :rtype: tuple of two :class:`Point`
+        """
+
+        w1 = max(map(len, col1.map))
+        h1 = len(col1.map)
+        tl1 = col1.pos
+        rb1 = tl1 + Point(w1, h1)
+
+        w2 = max(map(len, col2.map))
+        h2 = len(col2.map)
+        tl2 = col2.pos
+        rb2 = tl2 + Point(w2, h2)
+        if (tl1.x >= rb2.x or tl1.y >= rb2.y or
+                rb1.x <= tl2.x or rb1.y <= tl2.y):
+            # Definelty not overlapping
+            return
+        # Now find where exactly overlapping occured
+        tl_ovlp = Point(max(tl1.x, tl2.x), max(tl1.y, tl2.y))
+        rb_ovlp = Point(min(rb1.x, rb2.x), min(rb1.y, rb2.y))
+        # Now find if they actually collided
+        # first, calculate offsets
+        ox1 = abs(tl1.x - tl_ovlp.x)
+        oy1 = abs(tl1.y - tl_ovlp.y)
+        ox2 = abs(tl2.x - tl_ovlp.x)
+        oy2 = abs(tl2.y - tl_ovlp.y)
+        # iterate over overlapping region
+        # and search for collision
+        for i in range(rb_ovlp.x - tl_ovlp.x):
+            for j in range(rb_ovlp.y - tl_ovlp.y):
+                # TODO: check length of current y-level string
+                # it might be not enough to contain i + ox1/2 element
+                if (col1.map[j + oy1][i + ox1] ==
+                        col2.map[j + oy2][i + ox2] == "#"):
+                    return (tl_ovlp, rb_ovlp)
+
+    def _add_collision(self, collider, other_type, callback=None):
+        if collider not in self._colliders:
+            raise ValueError("Attempt to add collision handler for "
+                             "unregistered collider {0}".format(collider))
+        type_pair = TypePair(collider.type, other_type)
+        if type_pair not in self._collisions.keys():
```

Comment:
There's good thing for this.

```python
self._collisions.setdefault(type_pair, []).append(callback)
```

Here example:
```python
>>> a = {}
>>> a.setdefault("key1", []).append(5)
>>> a
{'key1': [5]}
>>> a.setdefault("key1", []).append(10)
>>> a
{'key1': [5, 10]}
```
If key is present - return value
if not - return default value ([])